Add doc for filter support for cli tools (#11012)
gravitational/teleport: docs/pages/setup/reference/cli.mdx

tctl version
```

> ## Resource Filtering

Reviewer: ```suggestion
## Resource filtering
```

tctl version
```

> ## Resource Filtering

Reviewer: Thanks a lot for putting this together! There are a lot of guides that hint at our filter syntax, but nothing so far that spells it out in a comprehensive reference.

Since our filter predicate syntax is used throughout Teleport, though, I do think the `### Predicate Language` section needs to be a separate reference guide in `docs/pages/setup/reference`. We can then link to that new page in this H2 section.

Do different parts of the predicate syntax apply to different situations in Teleport? I don't see the `contains` function in this reference, for example. If this is the case, we should separate the predicate reference guide into subsections that apply to these different situations.

I think a central, dedicated predicate reference will go a long way toward clarifying this syntax, which so far is only documented via examples here and there in our docs.

Author: created separate page for predicate langugae: https://github.com/gravitational/teleport/blob/lisa/update-cli-doc/docs/pages/setup/reference/predicate-language.mdx

> Both `tsh` and `tctl` allow you to filter node, app, db, and kube resources using the `--search` and `--query` flags.

Reviewer: ```suggestion
Both `tsh` and `tctl` allow you to filter `node`, `app`, `db`, and `kube` resources using the `--search` and `--query` flags.
```
An alternative would be to say "filter Nodes, applications, databases, and Kubernetes resources"

> ### tsh and tctl ls
> 
> Resources nodes, apps, db, and kube all use the same filter flags and argument.
> 
> #### Argument
> 
> `[<labels>]` List of comma separated labels to filter by labels (e.g. key1=value1,key2=value2)

Reviewer 2: I think there's too many headings here. Can we simplify? Maybe something like this:

```suggestion
Both `tsh` and `tctl` allow you to filter node, app, db, and kube resources using the `--search` and `--query` flags.

The `--search` flag performs a simple fuzzy search on resource fields. For example, `--search=mac` searches for resources containing `mac`.

The `--query` flag allows you to perform more sophisticated searches using a predicate language that is detailed in the next section.

In both cases, you can further refine the results by appending a list of comma-separated labels to the command. For example:

```
$ tsh ls --search=term foo=bar
```
```

```bash
$ tsh ls --search=foo,bar labelKey1=labelValue1,labelKey2=labelValue2
```

### [Predicate Language](#predicate-language)

> The `--query` flag uses the same predicate language used else where but tailored for resources. 

Reviewer 2: ```suggestion
The `--query` flag uses the same predicate language used elsewhere but tailored for resources. 
```

I'm curious if @ptgott has suggestions on improving this. I don't think we should expect readers to know where "elsewhere" is, but I'm drawing a blank on better options.

| !=       | not equal to                         | `labels.env != "prod"`                              |
| &&       | and (all conditions must match)      | `labels.env == "prod" && labels.os == "mac"`        |
| \|\|     | or  (any one condition should match) | `labels.env == "dev" \|\| labels.env == "qa"`       |
| !        | not (used with functions)            | `!equals(labels.env, "prod")`                       |

The language also supports the following functions:

| Functions (with examples)             | Description                                          |
|---------------------------------------|------------------------------------------------------|
| `equals(labels.env, "prod")`          | resources with label key `env` equal to `prod`       |
| `exists(labels.env)`                  | resources with a label key `env`; value unchecked    |
| `!exists(labels.env)`                 | resources without a label key `env`; value unchecked |
| `search("foo", "bar", "some phrase")` | fuzzy match against common resource fields           |

#### Examples

The example with filters, searches for nodes with labels with key `env` equal to `staging` and key `os` equal to `mac`.

> ```bash
> # List all nodes
> tsh ls
> # List nodes using label argument
> tsh ls env=staging,os=mac
> # List nodes using search keywords
> tsh ls --search=staging,mac
> # List nodes using predicate language
> tsh ls --query='labels.env == "staging" && equals(labels.os, "mac")'
> ```

Reviewer 2: ```suggestion
```bash
# List all nodes
$ tsh ls

# List nodes using label argument
$ tsh ls env=staging,os=mac

# List nodes using search keywords
$ tsh ls --search=staging,mac

# List nodes using predicate language
$ tsh ls --query='labels.env == "staging" && equals(labels.os, "mac")'
```
```
